core/embed/rust/src/ui/shape/unlock_overlay.rs

```diff
@@ -0,0 +1,116 @@
+/// The overlay covers area of 170x170 pixels (centered at `pos`).
+pub struct UnlockOverlay {
+    /// Center of the overlay
+    pos: Point,
+    // Angle of the rotation (in degrees)
+    angle: f32,
+}
+
+impl UnlockOverlay {
+    /// Outer radius
+    pub const RADIUS: i16 = 85;
+    /// Distance between the circles
+    pub const SPAN: i16 = 10;
+    /// Thickness of the circles
+    pub const THICKNESS: i16 = 4;
+
+    /// Create a new overlay with the given center and rotation angle.
+    pub fn new(pos: Point, angle: f32) -> Self {
+        Self { pos, angle }
+    }
+
+    pub fn render<'a>(self, renderer: &mut impl Renderer<'a>) {
+        renderer.render_shape(self);
+    }
+
+    fn prepare_overlay(&self, canvas: &mut dyn Canvas) {
+        let center = canvas.bounds().center();
+
+        let transp = Color::black();
+        let opaque = Color::white();
+
+        canvas.fill_background(opaque);
+
+        // The most outer circle (with two sectors)
+        let angle = self.angle;
+        let r = Self::RADIUS;
+        canvas.fill_sector(center, r, 0.0 + angle, 140.0 + angle, transp);
+        canvas.fill_sector(center, r, 235.0 + angle, 270.0 + angle, transp);
+        canvas.fill_circle(center, r - Self::THICKNESS, opaque, 255);
+
+        // The second circle (with one sector)
+        let angle = -self.angle - 20.0;
+        let r = Self::RADIUS - Self::SPAN;
+        canvas.fill_sector(center, r, 0.0 + angle, 135.0 + angle, transp);
+        canvas.fill_circle(center, r - Self::THICKNESS, opaque, 255);
+
+        // The third circle (with one sector)
+        let angle = self.angle / 2.0 + 90.0;
+        let r = Self::RADIUS - 2 * Self::SPAN;
+        canvas.fill_sector(center, r, 0.0 + angle, 270.0 + angle, transp);
+        canvas.fill_circle(center, r - Self::THICKNESS, opaque, 255);
+
+        // The fourth circle (with two sectors)
+        let angle = -self.angle / 2.0 + 60.0;
+        let r = Self::RADIUS - 3 * Self::SPAN;
+        canvas.fill_sector(center, r, 0.0 + angle, 110.0 + angle, transp);
+        canvas.fill_sector(center, r, 180.0 + angle, 280.0 + angle, transp);
+        canvas.fill_circle(center, r - Self::THICKNESS, opaque, 255);
+
+        // Innner fixed circle
+        let r = Self::RADIUS - (9 * Self::SPAN) / 2;
+        canvas.fill_circle(center, r, transp, 255);
+        canvas.fill_circle(center, r - Self::THICKNESS, opaque, 255);
+    }
+}
+
+impl<'a> Shape<'a> for UnlockOverlay {
+    fn bounds(&self) -> Rect {
+        Rect::new(
+            self.pos - Offset::uniform(Self::RADIUS),
+            self.pos + Offset::uniform(Self::RADIUS + 1),
+        )
+    }
+
+    fn cleanup(&mut self, _cache: &DrawingCache<'a>) {
+        // TODO: inform the cache that we won't use the zlib slot anymore
+    }
+
+    fn draw(&mut self, canvas: &mut dyn Canvas, cache: &DrawingCache<'a>) {
+        let bounds = self.bounds();
+
+        let overlay_buff = &mut unwrap!(cache.image_buff(), "No image buffer");
```

Comment:
maybe image_buff should be divided and we should use dedicated cache.overlay_buff for this (which would also allow cleaner disable of this feature on models with insufficient memory)

Reply:
`image_buff` in `DrawingCache` serves as a temporary buffer for shapes that can be used within its `draw()` method for any purpose. Having it as a single image is more universal than splitting it into two distinguished images.

But I absolutely agree that we need to come up with a better approach to memory management within the shape library.

Reply:
I was thinking that with the divided buffers we could use the overlay along with drawing stuff that needs the image_buff - thus it would be actually more useful having it divided. 

But we can leave it as is and solve later along with the memory management if you prefer, i will leave that up to you